### ISSUE TITLE
Is there a permanent way to set the internal log path dynamically (in code) on start up? Despite additional changes to the NLog config in code thereafter.

### ISSUE THREAD

**Issue by davrob01** [NLog/NLog#2124]
Having read this: https://github.com/NLog/NLog/wiki/Internal-logging
I have this following issue with setting the internal log file path in code on start up:

**Type** (choose one):

- Bug / Question

**NLog version**:  4.4.5

**Platform**: .Net 4.5

In case of a BUG:

- What is the current result? Changes to the internal LogFile location in code are not applied or do not persist.
- What is the expected result? Changes to the intneral LogFile location in code should be applied and persist.

- Are there any work arounds? yes

1. A work around is to Set the InternalLogger LogFile at some later point in the start up of the application (not sure yet at what point I am first allowed to make changes to the internal log file path in code, but it seems like I must do so after other configuration changes set in code). However, as I will explain, this will result in two internal log files created in total.

2. The other work around is to just not do this at all and manually set the path for each environment's  config and each application's config.

**Detailed Summary:**

So I have a Start method in its own class that I use to initialize any log related settings that must be set dynamically in code, rather than in the XML of the NLog.config. I am calling this from my Global.asax.cs's Application_Start()

One of the things I am also doing is in this Start method is registering a custom layout renderer. Since I realize that it must be done on application start.
This works perfectly fine.

Other things I am doing are setting Configuration variables, such as the log path that will be used by targets.
This works fine.

And I was hoping that, in this same Start method, I could also change the location of the internal log file. I would of course like to do this as soon as possible in the start up pipeline. So I thought I would do it first thing in my Start method. However, the changes do not seem to persist unless I perform them later in code.

So the work around is to perform the setting of the internal log at the end of my start method, which works but it leaves a mess. What I mean is:
(To verify when the internal log writes, I set it to Debug) What happens is, I get part of the output of the internal log in the original log file path from the config, and the rest in the new path I set dynamically.

This is not ideal.

Ideally, I would like to set the path dynamically as soon as possible so that everything is logged to the same place COMPLETELY.

So my question is:
**Is there a permanent way to set the internal log path dynamically on start up?**

in case of a FEATURE REQUEST:
 
 - Why do we need it?

I realize the internal log is mainly meant for debugging purposes and not necessarily mission critical, but it would be nice to have this feature anyway since our path to the logs is different in various environments and applications and of course there are some things that can go wrong internally so I am certainly using the internal log for at least Error level logging in my applications. I am glad I can at least accomplish the dynamic setting of paths for the actual target logs.

Attached is the code for my Start method, which I am calling from my Global.asax.cs's Application_Start()
Note: LogProvider class is simply a class I made that serves as a wrapper for a NLog.Logger instance

here is some of my nlog config when testing this, note the different path for the internal log, purposely used for testing if the dynamicly set path (set in my Start method) is used instead.
```
<nlog xmlns="http://www.nlog-project.org/schemas/NLog.xsd"
      xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
      xsi:schemaLocation="http://www.nlog-project.org/schemas/NLog.xsd NLog.xsd"
      autoReload="true"
      throwExceptions="false"
      throwConfigExceptions="false"
      internalLogLevel="Debug" internalLogFile="C:\applicationlogs\NavigationManager\nlog-internal.log">
```

[Start up.txt](https://github.com/NLog/NLog/files/1015397/Start.up.txt)

**Comment by davrob01** [comment]
Update:

I realized I can accomplish my desired result.

**The only thing I need to do is just not ever set the internalLogFile path in the xml config at all whatsoever.**

And ONLY set the path in code.

I see this as a pretty fair solution rather than a work around. I was just under the impression I could have a, sort of, default path in the xml config and set a path in code additionally. But based on all the other config changes I am doing, it appears that I must pick one way or the other for setting the internal log path. I am okay with this.

You may close this issue as you see fit.

Thank you

**Comment by 304NotModified** [comment]
> What is the expected result? Changes to the intneral LogFile location in code should be applied and persist.

You have to set it as soon as possible **and** after `LogManager.ConfigurationReloaded` (event)

**Comment by davrob01** [comment]
AFAIK it doesn't appear that I am doing anything in my Start method to trigger LogManager.ConfigurationReloaded
to occur. 
The path I set for the internal log is maintained, since it is still written to when I perform other actions in my application that write to logs.

Furthermore, I noticed that if I do an explicit change to the NLog.config's xml , such as switching the internal log level, then my Start method is called because a change to the config triggers Application_Start()

Therefore, I think I can rest assured that the internal log will be written to without doing anything more. Based on what I am doing at least.

I suppose, I could add a handler for LogManager.ConfigurationReloaded and add code to it to set the internal log path, to fully ensure the internal log path is being set no matter what. Based on my scenario however, it seems like I do not have to do that since I am seeing it working despite what I am doing in my start method.

Edit: I added the following with a breakpoint in the begining of my Start method, to see if I am ever triggering ConfigurationReloaded event. It appears I am not. I may leave this handler implemented though as a safe guard anyway.

```
EventHandler<NLog.Config.LoggingConfigurationReloadedEventArgs> handler = (sender, e) =>
             {
                // ensure path is set on reload
                NLog.Common.InternalLogger.LogFile = "C:\\applicationlogs\\log.txt";
                NLog.Common.InternalLogger.Info("testing message"); // note: this is the only line written to the above location.
             };

            LogManager.ConfigurationReloaded += handler;
```

Thank you for your help on this.
